core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -383,11 +384,16 @@ protected boolean booleanEquality(SqlCallBinding binding,
   }
 
   /**
-   * CASE and COALESCE type coercion, collect all the branches types including then
+   * CASE WHEN and COALESCE type coercion, collect all the branches types including then
    * operands and else operands to find a common type, then cast the operands to the common type
    * when needed.
    */
   @Override public boolean caseWhenCoercion(SqlCallBinding callBinding) {
+    if (!(callBinding.getCall() instanceof SqlCase)) {
+      // For sql statement like: `coalesce(a, b, c)`
```

Comment:
This is better. The JavaDoc suggests that this was the original intent of the code too, but maybe it was never realized.